### PR TITLE
curvefs/client: s3 adaptor unit test optimization

### LINKED ISSUE #1226: s3 adaptor unit test optimization
**Is your feature request related to a problem? (你需要的功能是否与某个问题有关?)**
The current s3 adaptor only has integration tests, no complete unit tests, and needs to be supplemented

**Describe the solution you'd like (描述你期望的解决方法)**

**Describe alternatives you've considered (描述你想到的折衷方案)**

**Additional context/screenshots (更多上下文/截图)**

### PR DESCRIPTION
**Pull request by baijiaruo**
### What problem does this PR solve?

Issue Number: close #1226  

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/test/client/chunk_cache_manager_test.cpp [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+        S3ClientAdaptorOption option;
+        option.blockSize = 1 * 1024 * 1024;
+        option.chunkSize = 4 * 1024 * 1024;
+        option.pageSize = 64 * 1024;
+        option.intervalSec = 5000;
+        option.flushIntervalSec = 5000;
+        option.readCacheMaxByte = 104857600;
+        option.writeCacheMaxByte = 10485760000;
+        option.diskCacheOpt.diskCacheType = (DiskCacheType)0;
+        s3ClientAdaptor_ = new S3ClientAdaptorImpl();
```
Comment (wu-hanqing) [inline curvefs/test/client/chunk_cache_manager_test.cpp:56]: undeleted

File: curvefs/test/client/chunk_cache_manager_test.cpp [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+ protected:
+    S3ClientAdaptorImpl *s3ClientAdaptor_;
+    std::shared_ptr<ChunkCacheManager> chunkCacheManager_;
+};
+
+TEST_F(ChunkCacheManagerTest, test_write_new_data) {
+    uint64_t offset = 0;
+    uint64_t len = 1024;
+    int length = len;
+    char *buf = new char[len];
```
Comment (wu-hanqing) [inline curvefs/test/client/chunk_cache_manager_test.cpp:76]: ditto

File: curvefs/test/client/chunk_cache_manager_test.cpp [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+
+    chunkCacheManager_->WriteNewDataCache(s3ClientAdaptor_, offset, len, buf);
+    ASSERT_EQ(65536, s3ClientAdaptor_->GetFsCacheManager()->GetDataCacheSize());
+}
+
+TEST_F(ChunkCacheManagerTest, test_add_read_data_cache) {
+    uint64_t offset = 0;
+    uint64_t len = 1024 * 1024;
```
Comment (wu-hanqing) [inline curvefs/test/client/chunk_cache_manager_test.cpp:84]: it's better to randomize these test parameters

File: curvefs/test/client/chunk_cache_manager_test.cpp [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+
+        S3ClientAdaptorOption option;
+        option.blockSize = 1 * 1024 * 1024;
+        option.chunkSize = 4 * 1024 * 1024;
+        option.pageSize = 64 * 1024;
+        option.intervalSec = 5000;
+        option.flushIntervalSec = 5000;
+        option.readCacheMaxByte = 104857600;
+        option.writeCacheMaxByte = 10485760000;
+        option.diskCacheOpt.diskCacheType = (DiskCacheType)0;
```
Comment (wu-hanqing) [inline curvefs/test/client/chunk_cache_manager_test.cpp:55]: it's better to test different parameter combinations

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 5]
```diff
@@ -0,0 +1,2910 @@
...
+    response.set_status(::curvefs::space::SpaceStatusCode::SPACE_OK);
+    response.set_chunkid(25);
+
+    EXPECT_CALL(mockSpaceAllocService_, AllocateS3Chunk(_, _, _, _))
+        .WillOnce(
+            DoAll(SetArgPointee<2>(response),
+                  Invoke(S3RpcService<AllocateS3ChunkRequest,
+                  AllocateS3ChunkResponse>)));
+    EXPECT_CALL(mockS3Client_, Upload(_, _, _))
+        .WillRepeatedly(Return(1 * 1024 * 1024));*/
```
Comment (wu-hanqing) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:234]: remove

**Review by wu-hanqing — COMMENTED** [review]
some resources are not deleted, you should recheck it.

File: curvefs/src/client/fuse_s3_client.cpp [hunk 6]
```diff
@@ -50,10 +50,24 @@ CURVEFS_ERROR FuseS3Client::Init(const FuseClientOption &option) {
     ::curvefs::client::common::S3Info2FsS3Option(s3Info, &fsS3Option);
     SetFuseClientS3Option(&opt, fsS3Option);
 
-    s3Client_ = std::make_shared<S3ClientImpl>();
-    s3Client_->Init(opt.s3Opt.s3AdaptrOpt);
-    ret = s3Adaptor_->Init(opt.s3Opt.s3ClientAdaptorOpt, s3Client_.get(),
-                           inodeManager_, mdsClient_);
+    auto s3Client = std::make_shared<S3ClientImpl>();
+    s3Client->Init(opt.s3Opt.s3AdaptrOpt);
+    if (opt.s3Opt.s3ClientAdaptorOpt.diskCacheOpt.diskCacheType !=
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_s3_client.cpp:55]: How to test this init function?

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 7]
```diff
@@ -34,9 +34,42 @@ namespace curvefs {
...
+    const S3ClientAdaptorOption &option, std::shared_ptr<S3Client> client,
+    std::shared_ptr<InodeCacheManager> inodeManager,
+    std::shared_ptr<MdsClient> mdsClient,
+    std::shared_ptr<DiskCacheManagerImpl> diskCacheManagerImpl) {
+    auto ret = Init(option, client, inodeManager, mdsClient);
+    if (ret != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "Init s3 client adaptor failed, ret: " << ret;
+        return ret;
+    }
+    if (HasDiskCache()) {
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:47]: Check whether diskCacheManagerImpl is null?

File: curvefs/src/client/s3/client_s3_adaptor.cpp [hunk 8]
```diff
@@ -73,33 +106,38 @@ S3ClientAdaptorImpl::Init(const S3ClientAdaptorOption &option, S3Client *client,
...
+    prefetchExecQueueNum_ = option.prefetchExecQueueNum;
+    diskCacheType_ = option.diskCacheOpt.diskCacheType;
+    memCacheNearfullRatio_ = option.nearfullRatio;
+    throttleBaseSleepUs_ = option.baseSleepUs;
+    flushIntervalSec_ = option.flushIntervalSec;
+    client_ = client;
+    inodeManager_ = inodeManager;
+    mdsClient_ = mdsClient;
+    fsCacheManager_ = fsCacheManager;
+    diskCacheManagerImpl_ = diskCacheManagerImpl;
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_adaptor.cpp:140]: Don't you need diskCacheManagerImpl to be null？

File: curvefs/src/client/s3/client_s3_adaptor.h [hunk 9]
```diff
@@ -65,10 +65,21 @@ class S3ClientAdaptor {
      * @brief Initailize s3 client
      * @param[in] options the options for s3 client
      */
-    virtual CURVEFS_ERROR Init(const S3ClientAdaptorOption& option,
-                               S3Client* client,
-                               std::shared_ptr<InodeCacheManager> inodeManager,
-                               std::shared_ptr<MdsClient> mdsClient) = 0;
+    virtual CURVEFS_ERROR
+    Init(const S3ClientAdaptorOption &option, std::shared_ptr<S3Client> client,
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_adaptor.h:69]: Why S3ClientAdaptorImpl need so many Init function?

File: curvefs/src/client/s3/client_s3_cache_manager.h [hunk 10]
```diff
@@ -210,32 +211,41 @@ class ChunkCacheManager
...
         return (dataWCacheMap_.empty() && dataRCacheMap_.empty());
     }
 
     virtual void ReleaseReadDataCache(uint64_t key);
-    void ReleaseCache();
+    virtual void ReleaseCache();
     void TruncateCache(uint64_t chunkPos);
     void UpdateWriteCacheMap(uint64_t oldChunkPos, DataCache *dataCache);
-
+    void AddWriteDataCacheForTest(DataCachePtr dataCache);
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_cache_manager.h:240]: Add
// for unit test 
for function

File: curvefs/src/client/s3/client_s3_cache_manager.h [hunk 11]
```diff
@@ -398,6 +416,15 @@ class FsCacheManager {
         return lruByte_;
     }
 
+    void SetFileCacheManagerForTest(uint64_t inodeId,
+                                    FileCacheManagerPtr fileCacheManager) {
+        WriteLockGuard writeLockGuard(rwLock_);
+
+        auto ret = fileCacheManagerMap_.emplace(inodeId, fileCacheManager);
+        assert(ret.second);
+        (void)ret;
```
Comment (ilixiaocui) [inline curvefs/src/client/s3/client_s3_cache_manager.h:425]: (void)ret is redundant

File: curvefs/test/client/chunk_cache_manager_test.cpp (hunk 4 above)
Comment (ilixiaocui) [reply curvefs/test/client/chunk_cache_manager_test.cpp:55]: agree

File: curvefs/test/client/chunk_cache_manager_test.cpp [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+
+    chunkCacheManager_->WriteNewDataCache(s3ClientAdaptor_, 1024, len,
+                                          dataCacheBuf);
+    chunkCacheManager_->ReadByWriteCache(offset, len, buf, 0, &requests);
+    ASSERT_EQ(1, requests.size());
+    requests.clear();
+    /*
+                 -----               ReadData
+                    ------           DataCache
+           */
```
Comment (ilixiaocui) [inline curvefs/test/client/chunk_cache_manager_test.cpp:150]: Why do you need three scenarios to test this type of use case？

File: curvefs/test/client/chunk_cache_manager_test.cpp [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+    auto dataCache = std::make_shared<DataCache>(
+        s3ClientAdaptor_, chunkCacheManager_, 1024, len, dataCacheBuf);
+    chunkCacheManager_->AddReadDataCache(dataCache);
+    chunkCacheManager_->ReadByReadCache(offset, len, buf, 0, &requests);
+    ASSERT_EQ(1, requests.size());
+    requests.clear();
+    /*
+                 -----               ReadData
+                    ------           DataCache
+           */
```
Comment (ilixiaocui) [inline curvefs/test/client/chunk_cache_manager_test.cpp:231]: ditto

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 14]
```diff
@@ -0,0 +1,2910 @@
...
+
+class ClientS3IntegrationTest : public testing::Test {
+ protected:
+    ClientS3IntegrationTest() {}
+    ~ClientS3IntegrationTest() {}
+    void SetUp() override {
+        Aws::InitAPI(awsOptions_);
+        ASSERT_EQ(0, server_.AddService(&mockMetaServerService_,
+                                        brpc::SERVER_DOESNT_OWN_SERVICE));
+        /*ASSERT_EQ(0, server_.AddService(&mockSpaceAllocService_,
```
Comment (ilixiaocui) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:105]: delete

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 15]
```diff
@@ -0,0 +1,2910 @@
...
+        ASSERT_EQ(0, server_.Start(addr_.c_str(), nullptr));
+        S3ClientAdaptorOption option;
+        option.blockSize = 1 * 1024 * 1024;
+        option.chunkSize = 4 * 1024 * 1024;
+        option.pageSize = 64 * 1024;
+        option.intervalSec = 5000;
+        option.flushIntervalSec = 5000;
+        option.readCacheMaxByte = 104857600;
+        option.writeCacheMaxByte = 10485760000;
+        // option.fuseMaxSize = 131072;
```
Comment (ilixiaocui) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:116]: ditto

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 16]
```diff
@@ -0,0 +1,2910 @@
...
+    ASSERT_EQ(1, fsCacheManager->GetDataCacheNum());
+
+    offset = 0;
+    s3ClientAdaptor_->Write(inode.inodeid(), offset, len, buf);
+    ASSERT_EQ(1, fsCacheManager->GetDataCacheNum());
+
+    offset = 2 * 1024 * 1024;
+    s3ClientAdaptor_->Write(inode.inodeid(), offset, len, buf);
+    ASSERT_EQ(1, fsCacheManager->GetDataCacheNum());
+    // ASSERT_EQ(0, s3ChunkInfoList.s3chunks(0).version());
```
Comment (ilixiaocui) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:206]: ditto

File: curvefs/test/client/client_s3_adaptor_Integration.cpp (hunk 5 above)
Comment (ilixiaocui) [reply curvefs/test/client/client_s3_adaptor_Integration.cpp:234]: ditto

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 17]
```diff
@@ -0,0 +1,2910 @@
...
+    auto inodeWrapper = std::make_shared<InodeWrapper>(inode, nullptr);
+    EXPECT_CALL(mockInodeManager_, GetInode(_, _))
+        .WillOnce(
+            DoAll(SetArgReferee<1>(inodeWrapper), Return(CURVEFS_ERROR::OK)))
+        .WillOnce(
+            DoAll(SetArgReferee<1>(inodeWrapper), Return(CURVEFS_ERROR::OK)));
+    s3ClientAdaptor_->Write(inode.inodeid(), offset, len, buf);
+    inode.set_length(offset + len);
+
+    // ASSERT_EQ(2, gObjectDataMaps.size());
```
Comment (ilixiaocui) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:677]: ditto

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 18]
```diff
@@ -0,0 +1,2910 @@
...
+
+TEST_F(ClientS3IntegrationTest, test_read_hole1) {
+    curvefs::metaserver::Inode inode;
+
+    InitInodeForIntegration(&inode);
+    uint64_t offset = 1 * 1024 * 1024;
+    uint64_t len = 1 * 1024 * 1024;
+    char *buf = new char[len];
+    memset(buf, 'a', len);
+    /*
```
Comment (ilixiaocui) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:654]: ditto. Check everything else. Delete everything that needs to be deleted

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 19]
```diff
@@ -0,0 +1,2910 @@
...
+    delete expectBuf;
+    std::map<std::string, S3Data>::iterator iter = gObjectDataMaps.begin();
+    for (; iter != gObjectDataMaps.end(); iter++) {
+        delete iter->second.buf;
+        iter->second.buf = NULL;
+    }
+    gObjectDataMaps.clear();
+}
+
+TEST_F(ClientS3IntegrationTest, test_read_hole1) {
```
Comment (ilixiaocui) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:646]: Some scenarios have several use cases. Note the differences.

File: curvefs/test/client/fs_cache_manager_test.cpp [hunk 20]
```diff
@@ -17,65 +17,118 @@
...
     std::list<DataCachePtr>::iterator outIter;
-    {
-        EXPECT_CALL(*mockCacheMgr, ReleaseReadDataCache(_)).Times(0);
 
-        for (size_t i = 0; i < maxReadCacheByte / smallDataCacheByte; ++i) {
-            manager.Set(std::make_shared<DataCache>(s3ClientAdaptor_,
-                                                    mockCacheMgr, 0,
-                                                    smallDataCacheByte, buf),
-                        &outIter);
+    {
```
Comment (ilixiaocui) [inline curvefs/test/client/fs_cache_manager_test.cpp:112]: Just to make a note of what each square bracket is testing.

File: curvefs/test/client/client_s3_adaptor_Integration.cpp (hunk 5 above)
Comment (baijiaruo) [reply curvefs/test/client/client_s3_adaptor_Integration.cpp]: fixed

File: curvefs/src/client/s3/client_s3_adaptor.h (hunk 9 above)
Comment (ilixiaocui) [reply curvefs/src/client/s3/client_s3_adaptor.h:69]: Distinguish whether diskcache is required by whether diskCacheManagerImpl is null or not

File: curvefs/test/client/chunk_cache_manager_test.cpp (hunk 12 above)
Comment (ilixiaocui) [reply curvefs/test/client/chunk_cache_manager_test.cpp:158]: Note what are these tests?

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 21]
```diff
@@ -0,0 +1,2849 @@
...
+    delete buf;
+    std::map<std::string, S3Data>::iterator iter = gObjectDataMaps.begin();
+    for (; iter != gObjectDataMaps.end(); iter++) {
+        delete iter->second.buf;
+        iter->second.buf = NULL;
+    }
+    gObjectDataMaps.clear();
+}
+
+TEST_F(ClientS3IntegrationTest, test_truncate_small2) {
```
Comment (ilixiaocui) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:994]: ditto. Note about the test scenario

File: curvefs/test/client/client_s3_adaptor_Integration.cpp [hunk 22]
```diff
@@ -0,0 +1,2849 @@
...
+    delete buf;
+    std::map<std::string, S3Data>::iterator iter = gObjectDataMaps.begin();
+    for (; iter != gObjectDataMaps.end(); iter++) {
+        delete iter->second.buf;
+        iter->second.buf = NULL;
+    }
+    gObjectDataMaps.clear();
+}
+
+TEST_F(ClientS3IntegrationTest, test_truncate_big1) {
```
Comment (ilixiaocui) [inline curvefs/test/client/client_s3_adaptor_Integration.cpp:1103]: ditto. Please note the following similar test cases.

File: curvefs/test/client/client_s3_adaptor_Integration.cpp (hunk 21 above)
Comment (baijiaruo) [reply curvefs/test/client/client_s3_adaptor_Integration.cpp:1061]: fixed

File: curvefs/test/client/client_s3_adaptor_Integration.cpp (hunk 22 above)
Comment (baijiaruo) [reply curvefs/test/client/client_s3_adaptor_Integration.cpp:1180]: fixed

**Review by wu-hanqing — APPROVED** [review]

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**